### PR TITLE
[release/7.0] Add query parameters to gRPC transcoding OpenAPI

### PR DESCRIPTION
**Pull request by JamesNK**
Fixes https://github.com/dotnet/aspnetcore/issues/42407

Changes:
* Fields that aren't bound with route parameters or body are added as query parameters.
* Change route parameters to use JSON name in pattern and parameter names.

Changes are all to a .NET 7 experimental package.

### CODE REVIEW

File: src/Grpc/JsonTranscoding/src/Shared/ServiceDescriptorHelpers.cs [hunk 1]
```diff
@@ -347,6 +348,88 @@ public static bool TryResolvePattern(HttpRule http, [NotNullWhen(true)]out strin
...
+        MethodDescriptor methodDescriptor,
+        MessageDescriptor? bodyDescriptor,
+        List<FieldDescriptor>? bodyFieldDescriptors)
+    {
+        var queryParameters = new Dictionary<string, FieldDescriptor>();
+        var existingParameters = new List<FieldDescriptor>();
+
+        if (routeParameters.Count > 0)
+        {
+            existingParameters.AddRange(routeParameters.Values.Last());
```
Comment (captainsafia) [inline src/Grpc/JsonTranscoding/src/Shared/ServiceDescriptorHelpers.cs:362]: Why the `Last` here and below?

File: src/Grpc/JsonTranscoding/src/Shared/ServiceDescriptorHelpers.cs [hunk 2]
```diff
@@ -347,6 +348,88 @@ public static bool TryResolvePattern(HttpRule http, [NotNullWhen(true)]out strin
...
+        if (bodyDescriptor != null)
+        {
+            if (bodyFieldDescriptors != null)
+            {
+                // Body with field name. Note: might not be a field on root request message.
+                existingParameters.Add(bodyFieldDescriptors.Last());
+            }
+            else
+            {
+                // Body with wildcard. All parameters at in the body.
```
Comment (captainsafia) [inline src/Grpc/JsonTranscoding/src/Shared/ServiceDescriptorHelpers.cs:374]:
```suggestion
                // Body with wildcard. All parameters are in the body.
```

File: src/Grpc/JsonTranscoding/src/Shared/ServiceDescriptorHelpers.cs [hunk 3]
```diff
@@ -347,6 +348,88 @@ public static bool TryResolvePattern(HttpRule http, [NotNullWhen(true)]out strin
...
+                    default:
+                        // Complex repeated fields aren't valid query parameters.
+                        if (!fieldDescriptor.IsRepeated)
+                        {
+                            RecursiveVisitMessages(queryParameters, existingParameters, fieldDescriptor.MessageType, path);
+                        }
+                        break;
+                }
+
+                path.RemoveAt(path.Count - 1);
```
Comment (captainsafia) [inline src/Grpc/JsonTranscoding/src/Shared/ServiceDescriptorHelpers.cs:428]: What's the purpose of removing this item?

File: src/Grpc/JsonTranscoding/src/Shared/ServiceDescriptorHelpers.cs (hunk 1 above)
Comment (JamesNK) [reply src/Grpc/JsonTranscoding/src/Shared/ServiceDescriptorHelpers.cs]: Added comment

File: src/Grpc/JsonTranscoding/src/Shared/ServiceDescriptorHelpers.cs (hunk 3 above)
Comment (JamesNK) [reply src/Grpc/JsonTranscoding/src/Shared/ServiceDescriptorHelpers.cs:436]: Added comment

**Review by captainsafia — APPROVED** [review]

**Comment by Pilchie** [comment]
Approved for .NET 7 RC2.

**Comment by JamesNK** [comment]
@dotnet/aspnet-build Merge please 🙏

**Pull request merged**